v3/mqtt.go

```diff
@@ -171,77 +221,74 @@ func NewMqttClient(mqttHost string, mqttPort int, tlsConfig *tls.Config,
 
 // Const function, but we don't want to pass client by value
 func (client *MqttClient) IsConnected() bool {
-	return client.isConnected
+	return client.conn.status == ConnectedNetworkStatus
+}
+
+// GetLastActivity will return the time since the last send or
+// receive.
+func (client *MqttClient) GetLastActivity() time.Time {
+	return client.conn.lastActivity
 }
 
 // exported for now, but pretty sure this needs to be called only from
 // NewMqttclient. This blocks until we get a Connection Ack
-func (client *MqttClient) Connect() error {
+func (client *MqttClient) Connect(ctx context.Context) error {
 	user, pwd := client.delegate.AuthInfo()
 	p := packet.NewConnectPacket()
 	p.Version = packet.Version311
 	p.Username = user
 	p.Password = pwd
 	p.CleanSession = true
 
-	respChan, err := client.conn.SendPacket(p)
+	respChan, err := client.conn.sendPacket(ctx, p)
 	if err != nil {
 		logError("[MQTT] failed to send packet: %s", err.Error())
 		return err
 	}
 
-	var pkt packet.Packet
 	client.wgRecv.Add(1)
-	if pkt, err = client.receivePacket(respChan); err != nil {
-		logError("[MQTT] failed to receive packet %s", err)
-		return err
+	resp := client.receivePacket(ctx, respChan)
+	if resp.Err != nil {
+		client.conn.status = DisconnectedNetworkStatus
+		return resp.Err
 	}
-
-	if pkt.Type() != packet.CONNACK {
-		logError("[MQTT] received unexpected packet %s", pkt.Type())
-		return errors.New("unexpected response")
-	}
-
-	ack := pkt.(*packet.ConnackPacket)
-	if ack.ReturnCode != packet.ConnectionAccepted {
-		return fmt.Errorf("connection failed: %s", ack.ReturnCode.Error())
-	}
-
```

Comment:
I moved most of the logic that works with the mqtt_packet into the connection - https://github.com/moderepo/device-sdk-go/blob/cf46cfcb74323ff788dda91a2630b3b9cffd4439/v3/mqtt.go#L571